Struct migration

MichalGniadek opened the pull request:
Hi, I saw that this it out of scope in the design doc, but I had already written it, so I figured I would a shoot you a PR anyway 😅

This PR implements simple struct migration when hot-patch happens that is able to handle struct renaming and field modifications. The same limitations apply as to function (struct needs to exist in the initial compilation).

User needs to derive `HotPatchMigrate` trait on a struct (and this struct must also implement `Component` and `Default`, and reflect all of these trait). Additionally, you need a `#[hot(rerun_on_hot_patch = true)]` that registers all the components. This can't be really removed until bevy implements some auto-registration solution.

The way this PR works is that it keeps a hashmap from `TypeId` to a hot function that returns `TypeId` for that type. When the struct is modified, the function start returning a different `TypeId` which can be detected and component from the previous `TypeId` can be migrated.

Note that docs are to do, it's completely unoptimized, and unwraps everywhere for now, because I wasn't sure if there's interest for this in the repo.

janhohenheim commented on examples/component_migration.rs:
```diff
@@ -0,0 +1,76 @@
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_plugins(SimpleSubsecondPlugin::default())
+        .add_systems(Startup, |mut commands: Commands| {
+            commands.spawn((
+                Example {
+                    field_a: 5,
+                    ..Default::default()
+                },
+                Example2 {
+                    field_b: 1005,
+                    ..Default::default()
+                },
+            ));
+            commands.spawn((
+                Example {
+                    field_a: 8,
+                    ..Default::default()
+                },
+                Example2 {
+                    field_b: 1008,
+                    ..Default::default()
+                },
+            ));
+        })
```
Could you move this into an fn? :)

janhohenheim commented on examples/component_migration.rs:
```diff
@@ -0,0 +1,76 @@
+        })
+        .add_systems(Startup, register_components)
+        .add_systems(Update, print_components)
+        .insert_resource(PrintTimer(Timer::from_seconds(1.0, TimerMode::Repeating)))
```
Instead of a timer, you can simply use `info_once!`

janhohenheim commented on macros/src/lib.rs:
```diff
@@ -287,3 +288,19 @@ fn is_result_unit(output: &ReturnType) -> bool {
     }
 }
+
+#[proc_macro_derive(HotPatchMigrate)]
```
TODO: documentation

MichalGniadek replied:
Also added `#![warn(missing_docs)]` and simple top-level and `#[hot]` docs. Let me know if that's not needed.

janhohenheim commented:
I'd really like to get this merged :) Would be happy if we could have someone like @bushrat011899 or @mrgvsv taking a look at the reflection side

MrGVSV commented on src/migration.rs:
```diff
@@ -0,0 +1,111 @@
+
+#[derive(Resource, Default, Deref, DerefMut)]
+pub(crate) struct ComponentMigrations(
+    HashMap<TypeId, Arc<dyn Fn() -> TypeId + Sync + Send + 'static>>,
```
I believe Bevy exports a `TypeIdMap` which is supposed to be a bit more optimized.

MrGVSV commented on src/migration.rs:
```diff
@@ -0,0 +1,111 @@
+                for e in es {
+                    match world.get_reflect(e, from) {
+                        Ok(c) => {
+                            let c = c.reflect_clone().unwrap();
```
Cloning isn't guaranteed to succeed or be implemented. Should we really be panicking here? Maybe we could print a warning and skip?

MrGVSV commented on macros/src/lib.rs:
```diff
@@ -287,3 +288,19 @@ fn is_result_unit(output: &ReturnType) -> bool {
+
+    let expanded = quote! {
+        impl bevy_simple_subsecond_system::migration::HotPatchMigrate for #name {
+            fn current_type_id() -> std::any::TypeId {
+                bevy_simple_subsecond_system::dioxus_devtools::subsecond::HotFn::current(|| std::any::TypeId::of::<Self>()).call(())
```
Nit: Haven't familiarized myself with this crate yet, but at least for Bevy, we prefer pulling from `core` so we're more no_std-ready

```suggestion
            fn current_type_id() -> core::any::TypeId {
                bevy_simple_subsecond_system::dioxus_devtools::subsecond::HotFn::current(|| core::any::TypeId::of::<Self>()).call(())
```

MichalGniadek replied:
Moved to core and bevy platform.

MrGVSV commented on src/migration.rs:
```diff
@@ -0,0 +1,111 @@
+    prelude::{Deref, DerefMut},
+    reflect::{FromType, Reflect, prelude::ReflectDefault},
+};
+use std::{
+    any::{Any, TypeId},
+    sync::Arc,
+};
```
Should we be using `bevy_platform` here? I don't know how no_std-compatible this crate or subsecond are

MichalGniadek replied:
Moved to core and bevy platform.

MrGVSV reviewed (commented):
Reflection usages make sense to me!

janhohenheim commented on examples/component_migration.rs:
```diff
@@ -0,0 +1,76 @@
+    registry.register::<Example2>();
+}
+
+// Try changing the components below:
```
```suggestion
// Try changing the components below at runtime:
```

janhohenheim commented on src/migration.rs:
```diff
@@ -0,0 +1,111 @@
+                                world.resource_scope::<AppTypeRegistry, ()>(|world, registry| {
+                                    let registry = registry.read();
+                                    let reflect_default =
+                                        registry.get_type_data::<ReflectDefault>(to).unwrap();
+                                    let reflect_component =
+                                        registry.get_type_data::<ReflectComponent>(to).unwrap();
```
Could we log an error instead of `unwrap`?

janhohenheim commented on src/migration.rs:
```diff
@@ -0,0 +1,111 @@
+                }
+            },
+        )
+        .unwrap();
```
Could we log an error instead of `unwrap`?

janhohenheim commented:
Note: the conflict should be simple to resolve, `main` now just gates the code behind a `#[cfg]`.

MichalGniadek commented:
Updated the `migrate_component` to be just better in general - more readable and only iterate over entities with the component that's being migrated.

MichalGniadek commented:
Assuming that users follow the suggestion of a hot, startup system that registers their components, we actually only need to run `register_migratable_components` just once in post startup (because unfortunately only components that existed in the initial compilation can be migrated).

MichalGniadek commented:
Huh, based on the error message it seems like it's trying to use the previous version of the component (with `Warrior`). Will take a look!

janhohenheim commented:
@MichalGniadek FYI the current main makes `#[hot]` no longer patch signatures by default. That's probably needed for your feature, so you may want to use `#[hot(hot_patch_signature = true)]` in the example :)

janhohenheim commented:
@MichalGniadek could we use a system set instead?

MichalGniadek commented:
Fair point. Added the migration system to a new system set (in `PreUpdate`) and updated the `#[hot]` macro to schedule its system `.before()` the set. Also updated the docs so now tests and docs CI should pass.

janhohenheim commented:
I meant a more classic enum holding the different "stages", but I can do that myself after merging, it's fine :)
